lib/amazonka-core/src/Amazonka/Error.hs

```diff
@@ -135,22 +137,23 @@ parseXMLError ::
   [Header] ->
   ByteStringLazy ->
   Error
-parseXMLError a s h bs = decodeError a s h bs (decodeXML bs >>= go)
+parseXMLError a s h bs = decodeError a s h bs (go <$> decodeXML bs)
   where
     go x =
       serviceError a s h
-        <$> code x
-        <*> may' (firstElement "Message" x)
-        <*> may' (firstElement "RequestId" x <|> firstElement "RequestID" x)
+      (code x)
+      (may' (firstElement "Message" x))
+      (may' (firstElement "RequestId" x) <|> may' (firstElement "RequestID" x))
 
-    code x =
-      Just <$> (firstElement "Code" x >>= parseXML)
-        <|> return root
+    code x
+      | Right y <- firstElement "Code" x >>= parseXML = y
+      | otherwise = root
```

Comment:
I find this style a little hard to read, I think because there's left-to-right data flow in the `firstElement "Code" x >>= parseXML` part, and also in the `= y` part, but it's interrupted by the pattern-guard, which is right-to-left. How about this?
```suggestion
    code x = fromRight root $ parseXML =<< firstElement "Code"
```
(`fromRight` is in `Data.Either` since `base-4.10.0.0`.)